src/matrix/storage/idb/utils.ts

```diff
@@ -112,33 +121,41 @@ export function txnAsPromise(txn) {
                 error = new StorageError(`Transaction on ${dbName} with stores ${storeNames} was aborted.`);
             }
             reject(error);
+            // @ts-ignore
             needsSyncPromise && Promise._flush && Promise._flush();
         });
     });
 }
 
-export function iterateCursor(cursorRequest, processValue) {
+type CursorIterator<T, I extends IDBCursor> = I extends IDBCursorWithValue ?
+    (value: T, key: IDBValidKey, cursor: IDBCursorWithValue) => { done: boolean, jumpTo?: IDBValidKey } :
+    (value: undefined, key: IDBValidKey, cursor: IDBCursor) => { done: boolean, jumpTo?: IDBValidKey }
+
+export function iterateCursor<T, I extends IDBCursor = IDBCursorWithValue>(cursorRequest: IDBRequest<I | null>, processValue: CursorIterator<T, I>): Promise<boolean> {
     // TODO: does cursor already have a value here??
-    return new Promise((resolve, reject) => {
+    return new Promise<boolean>((resolve, reject) => {
         cursorRequest.onerror = () => {
             reject(new IDBRequestError(cursorRequest));
+            // @ts-ignore
             needsSyncPromise && Promise._flush && Promise._flush();
         };
         // collect results
         cursorRequest.onsuccess = (event) => {
-            const cursor = event.target.result;
+            const cursor = (event.target as IDBRequest<I>).result;
             if (!cursor) {
                 resolve(false);
+                // @ts-ignore
                 needsSyncPromise && Promise._flush && Promise._flush();
                 return; // end of results
             }
-            const result = processValue(cursor.value, cursor.key, cursor);
+            const result = processValue(cursor["value"], cursor.key, cursor);
```

Comment:
I think that this may actually hurt intuition; it's not that we _know_ that the value is an `IDBCursorWithValue`, but that
we are okay with getting back an `undefined` if `cursor` is an `IDBCursor`.

Reply:
So had a look, and the idiomatic way to narrow the type seems to be [`in` narrowing](https://www.typescriptlang.org/docs/handbook/2/narrowing.html#the-in-operator-narrowing) here, which preserves type-safety without casts. Also see [this playground example](https://www.typescriptlang.org/play?#code/GYVwdgxgLglg9mABAQwM4GUpwE4FMAUAjlAFyICSAIgELlgAmuAHogD4U0DyARgFa7RMOXAEoyVaj36CseRAG8AUIhWIYwRPgBEcPgKhC8WtUmIiFy1VbxQQ2U1AB0u6Qdm4A3JZUBfb4hs7By8-RVBIWAQUDHcAJiJSDloGZjYkqX1DUXEuPRlhCyt1TWIAbR08t2EtAF1zJSsrYmdKrK8rP2tcW3tEYhDFIA).

Reply:
Wouldn't this introduce a condition / branch here where there previously wasn't one? Especially since the function call does not at all depend on whether `value` is present. If that's fine, my initial thought was `"value" in cursor ? cursor.value : undefined`, but that actually doesn't compile. 

It's not just a problem with ternary conditionals: seems like narrowing doesn't work in this case:
![2021-08-20-093609_1195x140_scrot](https://user-images.githubusercontent.com/4361282/130265626-b4c5f883-1064-4249-b361-54f23cdc83e4.png)

Reply:
This is likely because `cursor` does not have a union type. I personally still think that this (admittedly convoluted) implementation of `iterateCursor` is nicer than the alternative:

We could remove the `I` generic here and just accept a function argument that accepts `T | undefined`. The problem is that in this case every call site of `iterateCursor` needs to provide a function / lambda where the argument may be `undefined`. This in turn requires lots of extra tiptoeing around non-existent `undefined`s. 

The current approach (and I'll add a comment for this since the type of `CursorIterator` is indeed a whopper) makes for very clean usage of the API: whenever we pass in an `IDBRequest<IDBCursor>`, we have to pass in a lambda that accepts `undefined` and `IDBCursor`; on the other hand, whenever we pass in an `IDBRequest<IDBCursorWithValue>`, we have to pass in a lambda that accepts `T` (and only `T`) and an `IDBCursorWithValue`. There's thus no need for unnecessary `undefined` checks.

Reply:
I wrote up a comment explaining this in 50b7a8a3fd09e879777be4d0989c0378fb8baee7.